File: modules/signal-protocol/ios/Stores/PersistentKyberPreKeyStore.swift
//
//  PersistentKyberPreKeyStore.swift
//  Pods
//
//  TilliT Native — Signal Protocol bindings
//


import Foundation
import LibSignalClient

class PersistentKyberPreKeyStore: KyberPreKeyStore {
    // KyberPreKeys are LOCAL, used by ANY remote user, so use fixed prefix
    private let keyPrefix = "local"
    private let keychain = KeychainHelper.shared

    init(directoryURL: URL) {
        // directoryURL is ignored - kyberPreKeys are global/local, not per-session
    }

    func loadKyberPreKey(id: UInt32, context: StoreContext) throws -> KyberPreKeyRecord {
        let key = kyberPreKeyKey(for: id)
        guard let data = keychain.load(for: key) else {
            throw SignalError.invalidKey("No Kyber pre-key with id \(id)")
        }
        return try KyberPreKeyRecord(bytes: [UInt8](data))
    }

    func storeKyberPreKey(_ record: KyberPreKeyRecord, id: UInt32, context: StoreContext) throws {
        let key = kyberPreKeyKey(for: id)
        let data = Data(record.serialize())
        let saved = keychain.save(data: data, for: key)
        if !saved {
            throw SignalError.internalError("Failed to persist kyberPreKey to Keychain")
        }
    }

    func markKyberPreKeyUsed(id: UInt32, signedPreKeyId: UInt32, baseKey: PublicKey, context: any StoreContext) throws {
        // Mark the Kyber pre-key as used
        let key = usedKyberPreKeyKey(for: id)

        // Store metadata about the usage
        let metadata: [String: Any] = [
            "kyberPreKeyId": id,
            "signedPreKeyId": signedPreKeyId,
            "baseKeyData": Data(baseKey.serialize()).base64EncodedString(),
            "timestamp": Date().timeIntervalSince1970
        ]

        let data = try JSONSerialization.data(withJSONObject: metadata, options: [])
        let saved = keychain.save(data: data, for: key)
        if !saved {
            throw SignalError.internalError("Failed to persist kyberPreKey usage to Keychain")
        }
    }

    private func kyberPreKeyKey(for id: UInt32) -> String {
        return "kyberPreKey-\(keyPrefix)-\(id)"
    }

    private func usedKyberPreKeyKey(for id: UInt32) -> String {
        return "usedKyberPreKey-\(keyPrefix)-\(id)"
    }
}

File: modules/signal-protocol/android/src/main/java/expo/modules/signalprotocol/stores/PersistentKyberPreKeyStore.kt
package expo.modules.signalprotocol.stores

import android.content.Context
import org.signal.libsignal.protocol.InvalidKeyIdException
import org.signal.libsignal.protocol.ReusedBaseKeyException
import org.signal.libsignal.protocol.ecc.ECPublicKey
import org.signal.libsignal.protocol.state.KyberPreKeyRecord
import org.signal.libsignal.protocol.state.KyberPreKeyStore

class PersistentKyberPreKeyStore(
    keyPrefix: String,  // ignored - kyberPreKeys are global/local, not per-session
    context: Context
) : KyberPreKeyStore {
    // KyberPreKeys are LOCAL, used by ANY remote user, so use fixed prefix
    private val keyPrefix = "local"
    private val keystore = KeystoreHelper.getInstance(context)

    override fun loadKyberPreKey(id: Int): KyberPreKeyRecord {
        val key = kyberPreKeyKey(id)
        val data = keystore.load(key) ?: throw InvalidKeyIdException("No kyber pre-key with id $id")
        return KyberPreKeyRecord(data)
    }

    override fun loadKyberPreKeys(): MutableList<KyberPreKeyRecord> {
        val prefix = "kyberPreKey-$keyPrefix-"
        return keystore.getAllKeys()
            .filter { it.startsWith(prefix) }
            .mapNotNull { key ->
                keystore.load(key)?.let { KyberPreKeyRecord(it) }
            }
            .toMutableList()
    }

    override fun storeKyberPreKey(id: Int, record: KyberPreKeyRecord) {
        val key = kyberPreKeyKey(id)
        val saved = keystore.save(key, record.serialize())
        if (!saved) {
            throw RuntimeException("Failed to persist kyberPreKey to EncryptedSharedPreferences")
        }
    }

    override fun containsKyberPreKey(id: Int): Boolean {
        val key = kyberPreKeyKey(id)
        return keystore.exists(key)
    }

    override fun markKyberPreKeyUsed(kyberPreKeyId: Int, messageId: Int, baseKey: ECPublicKey) {
        val usedKey = kyberPreKeyUsedKey(kyberPreKeyId)
        val previousBaseKey = keystore.load(usedKey)

        if (previousBaseKey != null) {
            // This pre-key was already used — check if same baseKey (retry) or different (replay attack)
            if (!previousBaseKey.contentEquals(baseKey.serialize())) {
                throw ReusedBaseKeyException("Kyber pre-key $kyberPreKeyId reused with different base key")
            }
            // Same baseKey = legitimate retry, allow it
            return
        }

        // Record which baseKey consumed this pre-key, then delete the pre-key
        keystore.save(usedKey, baseKey.serialize())
        val key = kyberPreKeyKey(kyberPreKeyId)
        keystore.delete(key)
    }

    private fun kyberPreKeyKey(id: Int): String {
        return "kyberPreKey-$keyPrefix-$id"
    }

    private fun kyberPreKeyUsedKey(id: Int): String {
        return "kyberPreKeyUsed-$keyPrefix-$id"
    }
}
